### ISSUE TITLE
Component API

### ISSUE THREAD

**Issue by nandorojo** [marklawlor/tailwindcss-react-native#27]
Would you consider creating a factory function / export components from this library rather than the Babel magic? That would make me much more likely to use it. If anything, the Babel plug-in could be an optional optimization that happens inside this library's components, although I would hope it isn't necessary. 

I stayed away from Tamagui for the same reason. Requiring plugins makes debugging a bit of a black box with race conditions and forces me to rely a lot on a maintainer. I prefer to import things directly with predictable behavior.

If this isn't of interest to you, no worries! It's your project after all. But I think it could help increase adoption a lot (and perhaps simplify the code.) Curious to get your thoughts @marklawlor

**Comment by nandorojo** [comment]
For example, not being able to use className in loops and things like that are pretty big restrictions. Perhaps there is a solution down the line, but it would be much better imo to be able to do that in general.

Love the direction of the library so I'm hoping there's a fit for me!

**Comment by marklawlor** [comment]
Yes that's currently on the radar but I need to do some ground work first :smile: 

I'm currently working on an API that allows users to use the `useTailwind` hook explicitly.

```JSX
import { MotiView } from "moti";
import { TailwindProvider, useTailwind } from "tailwindcss-react-native";

export function Test() {
  return (
    <TailwindProvider>
      <MotiView
        from={useTailwind('opacity-0')}
        animate={useTailwind('opacity-1')}
        exit={useTailwind('opacity-0')}
       />
    </TailwindProvider>
  );
}
```

While that's probably not the best example :smile:  It shows how the library will work for components that require styles in situations where the className transform won't work; or for developers who don't want to use the className transform.

The `useTailwind` hook is also great for debugging as you can simply `console.log` the result. It also gives full control to the developer when and where the styles are compiled.

Once this hook has been released I'm going to expose a `styled` function to create styled components. These components would work as expected within loops and don't need to follow the rules of hooks.

```JSX
import { styled } from "tailwindcss-react-native"

const StyledView = styled(View)

<StyledView className="text-lg" />
```

While these APIs make the library more robust - they still require the babel plugin, as its purpose is to provide a streamline DevUX. Its primary use is to 

1. watch for file changes
2. run `tailwindcss` on the changed files
3. Injects the changed styles in a manner that works with `hot-module-reload`.

Without this plugin, I'd have to make compromises (eg `tailwind-rn` makes the user run background processes) on the DevUX.

**Comment by marklawlor** [comment]
Released in `v0.1.1`

@nandorojo It is now possible to use the library without babel. You will need to follow the [without babel](https://github.com/marklawlor/tailwindcss-react-native#additional-setup) steps in the README.

@jmeistrich This *should* be a non-breaking change for you, but let me know if you have any issues.

**Comment by marklawlor** [comment]
As an additional note: The babel plugin now also uses the Component API - removing the previous restrictions it had (rule of hooks, etc)

**Comment by marklawlor** [comment]
I'm going to track this under two separate issues

#31 - Ability to run the babel plugin for style injection only (no transformation)
#8 - Option to customise the className attribute

#31 Is the solution you need instead of using `allowedModules`. Having this option will create the best experience for people using `NextJS` & `expo`. I'll update my `solito` example when this is merged.

**Comment by marklawlor** [comment]
@nandorojo Those two issues have been released in `0.1.7`. I have updated https://github.com/marklawlor/solito-tailwind showing  one possible setup for the library to integrate into a solito mono-repo.

I believe this repo has the functionality your asking for

- No babel transformation
- Uses NextJS to compile CSS when running web
- Uses babel to compile styles when running native.

**Comment by jmeistrich** [comment]
@marklawlor It still works for us. Thanks for checking!
